From e9f52b45e92826638b6b3ba2e0c918b68f712bb5 Mon Sep 17 00:00:00 2001
From: Enrico Seiler <enrico.seiler@hotmail.de>
Date: Mon, 13 Jun 2022 16:47:10 +0200
Subject: [PATCH 06/12] [NOAPI] [DETAIL] Include md instead of hpp

---
 test/snippet/CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/test/snippet/CMakeLists.txt b/test/snippet/CMakeLists.txt
index 72c07ae1a..d5fde2761 100644
--- a/test/snippet/CMakeLists.txt
+++ b/test/snippet/CMakeLists.txt
@@ -50,7 +50,7 @@ foreach (source_snippet IN ITEMS "test/snippet/alphabet/nucleotide/@target_alpha
                                  "test/snippet/alphabet/nucleotide/@target_alphabet@_implicit_conversion_from_@source_alphabet@_inherit.cpp.in"
                                  "test/snippet/alphabet/nucleotide/@target_alphabet@_implicit_conversion_from_@source_alphabet@_vector.cpp.in"
                                  "test/snippet/alphabet/nucleotide/@target_alphabet@_implicit_conversion_from_@source_alphabet@_views.cpp.in"
-                                 "doc/fragments/@target_alphabet@_implicit_conversion_from_@source_alphabet@.hpp.in")
+                                 "doc/fragments/@target_alphabet@_implicit_conversion_from_@source_alphabet@.md.in")
     seqan3_generate_snippet("${source_snippet}" -Dtarget_alphabet=dna4 -Dsource_alphabet=rna4)
     seqan3_generate_snippet("${source_snippet}" -Dtarget_alphabet=rna4 -Dsource_alphabet=dna4)
     seqan3_generate_snippet("${source_snippet}" -Dtarget_alphabet=dna5 -Dsource_alphabet=rna5)
-- 
2.34.1

