From 9b7a4ac34e684797dea84820fa9d806078532eb2 Mon Sep 17 00:00:00 2001
From: marehr <marehr-github@marehr.dialup.fu-berlin.de>
Date: Wed, 17 Feb 2021 20:05:23 +0100
Subject: [PATCH] [NOAPI] Disable -Werror

---
 test/seqan3-test.cmake | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/test/seqan3-test.cmake b/test/seqan3-test.cmake
index 2c566cc80..36ff249c8 100644
--- a/test/seqan3-test.cmake
+++ b/test/seqan3-test.cmake
@@ -63,7 +63,7 @@ file(MAKE_DIRECTORY ${SEQAN3_TEST_CLONE_DIR}/googletest/include/)
 # seqan3::test exposes a base set of required flags, includes, definitions and
 # libraries which are in common for **all** seqan3 tests
 add_library (seqan3_test INTERFACE)
-target_compile_options (seqan3_test INTERFACE "-pedantic"  "-Wall" "-Wextra" "-Werror")
+target_compile_options (seqan3_test INTERFACE "-pedantic"  "-Wall" "-Wextra")
 target_link_libraries (seqan3_test INTERFACE "seqan3::seqan3" "pthread")
 target_include_directories (seqan3_test INTERFACE "${SEQAN3_TEST_INCLUDE_DIR}")
 add_library (seqan3::test ALIAS seqan3_test)
-- 
2.31.1

